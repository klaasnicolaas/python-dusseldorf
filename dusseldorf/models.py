"""Models for Open Data Platform of Dusseldorf."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass
class ParkAndRide:
    """Object representing a ParkAndRide."""

    entry_id: int
    name: str
    address: str
    district: int
    neighbourhood: str
    public_transport: str
    longitude: float
    latitude: float

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> ParkAndRide:
        """Return a ParkAndRide object from a dictionary.

        Args:
            data: The data from the API.

        Returns:
            A ParkAndRide object.
        """
        return cls(
            entry_id=int(data["entry_id"]),
            name=data["name"],
            address=f"{data.get('strasse')} {data.get('hausnummer')}{data.get('hs_zusatz')}",
            district=int(data["stadtbezirk"]),
            neighbourhood=str(data["stadt"][11:]),
            public_transport=data["nahverkehr"],
            longitude=float(data["longitude"]),
            latitude=float(data["latitude"]),
        )


@dataclass
class DisabledParking:
    """Object representing a DisabledParking."""

    entry_id: int
    name: str
    number: int
    address: str
    time_limit: str | None
    note: str | None
    longitude: float
    latitude: float

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> DisabledParking:
        """Return a DisabledParking object from a dictionary.

        Args:
            data: The data from the API.

        Returns:
            A DisabledParking object.
        """
        return cls(
            entry_id=int(data["entry_id"]),
            name=data["name"],
            number=int(data["anzahl"]),
            address=f"{data.get('strasse')} {data.get('hausnr')}",
            time_limit=data.get("zeitbegrenzung") or None,
            note=data.get("bemerkung") or None,
            longitude=float(data["longitude"]),
            latitude=float(data["latitude"]),
        )


@dataclass
class Garage:
    """Object representing a Garage."""

    entry_id: int
    name: str
    address: str
    location: str
    longitude: float
    latitude: float

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> Garage:
        """Return a Garage object from a dictionary.

        Args:
            data: The data from the API.

        Returns:
            A Garage object.
        """
        return cls(
            entry_id=int(data["entry_id"]),
            name=data["name"],
            address=data["adresse_zufahrt"],
            location=data["ort"],
            longitude=float(data["longitude"]),
            latitude=float(data["latitude"]),
        )
