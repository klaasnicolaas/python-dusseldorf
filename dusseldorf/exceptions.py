"""Asynchronous Python client providing Open Data information of Dusseldorf."""


class ODPDusseldorfError(Exception):
    """Generic Open Data Platform Dusseldorf exception."""


class ODPDusseldorfConnectionError(ODPDusseldorfError):
    """Open Data Platform Dusseldorf - connection error."""
