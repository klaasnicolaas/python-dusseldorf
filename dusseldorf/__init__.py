"""Asynchronous Python client providing Open Data information of Dusseldorf."""

from .dusseldorf import ODPDusseldorf
from .exceptions import ODPDusseldorfConnectionError, ODPDusseldorfError
from .models import DisabledParking, Garage, ParkAndRide

__all__ = [
    "ODPDusseldorf",
    "ODPDusseldorfConnectionError",
    "ODPDusseldorfError",
    "Garage",
    "ParkAndRide",
    "DisabledParking",
]
