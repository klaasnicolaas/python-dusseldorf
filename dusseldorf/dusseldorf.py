"""Asynchronous Python client providing Open Data information of Dusseldorf."""
from __future__ import annotations

import asyncio
import socket
from dataclasses import dataclass
from importlib import metadata
from typing import Any

import aiohttp
import async_timeout
from aiohttp import hdrs
from yarl import URL

from .exceptions import ODPDusseldorfConnectionError, ODPDusseldorfError
from .models import DisabledParking, Garage, ParkAndRide


@dataclass
class ODPDusseldorf:
    """Main class for handling data fetchting from Open Data Platform of Dusseldorf."""

    request_timeout: float = 10.0
    session: aiohttp.client.ClientSession | None = None

    _close_session: bool = False

    async def _request(
        self,
        uri: str,
        *,
        method: str = hdrs.METH_GET,
        params: dict[str, Any] | None = None,
    ) -> Any:
        """Handle a request to the Open Data Platform API of Dusseldorf.

        Args:
            uri: Request URI, without '/', for example, 'status'
            method: HTTP method to use, for example, 'GET'
            params: Extra options to improve or limit the response.

        Returns:
            A Python dictionary (text) with the response from
            the Open Data Platform API of Dusseldorf.

        Raises:
            ODPDusseldorfConnectionError: Timeout occurred while
                connecting to the Open Data Platform API.
            ODPDusseldorfError: If the data is not valid.
        """
        version = metadata.version(__package__)
        url = URL.build(
            scheme="https",
            host="opendata.duesseldorf.de",
            path="/api/action/datastore/",
        ).join(URL(uri))

        headers = {
            "Accept": "application/json",
            "User-Agent": f"PythonODPDusseldorf/{version}",
        }

        if self.session is None:
            self.session = aiohttp.ClientSession()
            self._close_session = True

        try:
            async with async_timeout.timeout(self.request_timeout):
                response = await self.session.request(
                    method,
                    url,
                    params=params,
                    headers=headers,
                    ssl=True,
                )
                response.raise_for_status()
        except asyncio.TimeoutError as exception:
            raise ODPDusseldorfConnectionError(
                "Timeout occurred while connecting to the Open Data Platform API."
            ) from exception
        except (aiohttp.ClientError, socket.gaierror) as exception:
            raise ODPDusseldorfConnectionError(
                "Error occurred while communicating with Open Data Platform API."
            ) from exception

        content_type = response.headers.get("Content-Type", "")
        if "application/json" not in content_type:
            text = await response.text()
            raise ODPDusseldorfError(
                "Unexpected content type response from the Open Data Platform API",
                {"Content-Type": content_type, "Response": text},
            )

        return await response.json()

    async def disabled_parkings(self, limit: int = 10) -> list[DisabledParking]:
        """Get list of disabled parkings.

        Args:
            limit: Maximum number of disabled parkings to return.

        Returns:
            A list of disabled parking objects.
        """
        results: list[DisabledParking] = []
        locations = await self._request(
            "search.json",
            params={
                "resource_id": "995914c7-4275-49c4-8441-426722336c3a",
                "limit": limit,
            },
        )
        for item in locations["result"]["records"]:
            results.append(DisabledParking.from_dict(item))
        return results

    async def garages(self, limit: int = 10) -> list[Garage]:
        """Get list of garages.

        Args:
            limit: Maximum number of garages to return.

        Returns:
            A list of garage objects.
        """
        results: list[Garage] = []
        locations = await self._request(
            "search.json",
            params={
                "resource_id": "53d63e70-0ed4-4175-9924-c45530d5bf29",
                "limit": limit,
            },
        )
        for item in locations["result"]["records"]:
            results.append(Garage.from_dict(item))
        return results

    async def park_and_rides(self, limit: int = 10) -> list[ParkAndRide]:
        """Get list of park and rides.

        Args:
            limit: Maximum number of park and rides to return.

        Returns:
            A list of park and ride objects.
        """
        results: list[ParkAndRide] = []
        locations = await self._request(
            "search.json",
            params={
                "resource_id": "1fd5b3f3-ea03-4922-9fae-3577fbc9f78a",
                "limit": limit,
            },
        )
        for item in locations["result"]["records"]:
            results.append(ParkAndRide.from_dict(item))
        return results

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> ODPDusseldorf:
        """Async enter.

        Returns:
            The Open Data Platform Dusseldorf object.
        """
        return self

    async def __aexit__(self, *_exc_info: Any) -> None:
        """Async exit.

        Args:
            _exc_info: Exec type.
        """
        await self.close()
