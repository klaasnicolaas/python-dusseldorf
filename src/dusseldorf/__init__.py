"""Asynchronous Python client providing Open Data information of Dusseldorf."""

from .dusseldorf import ODPDusseldorf
from .exceptions import ODPDusseldorfConnectionError, ODPDusseldorfError
from .models import DisabledParking, Garage, ParkAndRide

__all__ = [
    "DisabledParking",
    "Garage",
    "ODPDusseldorf",
    "ODPDusseldorfConnectionError",
    "ODPDusseldorfError",
    "ParkAndRide",
]
