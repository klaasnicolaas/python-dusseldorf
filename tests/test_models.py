"""Test the models."""
import aiohttp
import pytest
from aresponses import ResponsesMockServer

from dusseldorf import DisabledParking, Garage, ODPDusseldorf, ParkAndRide

from . import load_fixtures


@pytest.mark.asyncio
async def test_all_garages(aresponses: ResponsesMockServer) -> None:
    """Test all garages function."""
    aresponses.add(
        "opendata.duesseldorf.de",
        "/api/action/datastore/search.json",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("garages.json"),
        ),
    )
    async with aiohttp.ClientSession() as session:
        client = ODPDusseldorf(session=session)
        spaces: list[Garage] = await client.garages()
        assert spaces is not None
        for item in spaces:
            assert isinstance(item, Garage)
            assert isinstance(item.entry_id, int)
            assert item.entry_id is not None
            assert isinstance(item.name, str)
            assert isinstance(item.longitude, float)
            assert isinstance(item.latitude, float)


@pytest.mark.asyncio
async def test_all_park_and_rides(aresponses: ResponsesMockServer) -> None:
    """Test park and ride spaces function."""
    aresponses.add(
        "opendata.duesseldorf.de",
        "/api/action/datastore/search.json",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("park_and_rides.json"),
        ),
    )
    async with aiohttp.ClientSession() as session:
        client = ODPDusseldorf(session=session)
        spaces: list[ParkAndRide] = await client.park_and_rides()
        assert spaces is not None
        for item in spaces:
            assert isinstance(item, ParkAndRide)
            assert isinstance(item.entry_id, int)
            assert item.entry_id is not None
            assert isinstance(item.name, str)
            assert item.district is not None
            assert item.neighbourhood is not None
            assert isinstance(item.longitude, float)
            assert isinstance(item.latitude, float)


@pytest.mark.asyncio
async def test_all_disabled_parkings(aresponses: ResponsesMockServer) -> None:
    """Test park and ride spaces function."""
    aresponses.add(
        "opendata.duesseldorf.de",
        "/api/action/datastore/search.json",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("disabled_parkings.json"),
        ),
    )
    async with aiohttp.ClientSession() as session:
        client = ODPDusseldorf(session=session)
        spaces: list[DisabledParking] = await client.disabled_parkings()
        assert spaces is not None
        for item in spaces:
            assert isinstance(item, DisabledParking)
            assert isinstance(item.entry_id, int)
            assert item.entry_id is not None
            assert isinstance(item.name, str)
            assert item.number is not None
            assert isinstance(item.longitude, float)
            assert isinstance(item.latitude, float)
