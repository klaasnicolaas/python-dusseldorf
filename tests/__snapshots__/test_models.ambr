# serializer version: 1
# name: test_all_disabled_parkings
  list([
    DisabledParking(entry_id='a29702e7-94ae-4556-8e83-b57e8368f7c9', number=4, address='Paul-Thomas-Straße 35', district='Holthausen', time_limit=None, note='Stadtbad Niederheid', longitude=6.85161382, latitude=51.17041525, last_update=datetime.datetime(2022, 10, 28, 0, 0, tzinfo=datetime.timezone.utc)),
    DisabledParking(entry_id='eec91bb1-95ae-4be4-8c10-c34e95b1a9a7', number=2, address='Peter-Behrens-Straße 2', district='Garath', time_limit=None, note='Einkaufszentrum', longitude=6.88558867, latitude=51.14448552, last_update=datetime.datetime(2022, 10, 28, 0, 0, tzinfo=datetime.timezone.utc)),
    DisabledParking(entry_id='e9616d07-69a2-4e1c-9efd-48ddb0351dfd', number=1, address='Pionierstraße 1', district='Friedrichstadt', time_limit=None, note='Arztpraxis', longitude=6.78565818, latitude=51.21858693, last_update=datetime.datetime(2022, 10, 28, 0, 0, tzinfo=datetime.timezone.utc)),
    DisabledParking(entry_id='119bc152-9ace-4f7a-8228-7fefe3c6f791', number=1, address='Pöhlenweg 41', district='Grafenberg', time_limit=None, note='mehrere Berechtigte/Wohngebiet', longitude=6.83597885, latitude=51.23818121, last_update=datetime.datetime(2022, 10, 28, 0, 0, tzinfo=datetime.timezone.utc)),
    DisabledParking(entry_id='a9c27189-1c56-4520-9492-582bcfc22b31', number=2, address='Posener Straße', district='Lierenfeld', time_limit=None, note='Kirche', longitude=6.83048486, latitude=51.20878751, last_update=datetime.datetime(2022, 10, 28, 0, 0, tzinfo=datetime.timezone.utc)),
    DisabledParking(entry_id='4188a370-9bb6-4dbf-8daf-1bd464d9b9a3', number=1, address='Poststraße 1', district='Carlstadt', time_limit=None, note='Maxkirche / Carlsplatz', longitude=6.77169835, latitude=51.22297154, last_update=datetime.datetime(2022, 10, 28, 0, 0, tzinfo=datetime.timezone.utc)),
    DisabledParking(entry_id='1fa87962-5cd3-4442-b148-85fa6d87693c', number=1, address='Potsdamer Straße 39', district='Hassels', time_limit='Mo.-So. 8:00-18:00', note='DRK', longitude=6.86189224, latitude=51.18448969, last_update=datetime.datetime(2022, 10, 28, 0, 0, tzinfo=datetime.timezone.utc)),
    DisabledParking(entry_id='c4275963-5d4f-4208-ab12-2da7f501c115', number=1, address='Prinzenallee 19', district='Heerdt', time_limit=None, note='Ärztehaus', longitude=6.7345448, latitude=51.23699852, last_update=datetime.datetime(2022, 10, 28, 0, 0, tzinfo=datetime.timezone.utc)),
    DisabledParking(entry_id='595db356-0e41-4014-8702-6e8d975f0a95', number=2, address='Quadenhofstraße', district='Gerresheim', time_limit=None, note='Friedhof', longitude=6.86591546, latitude=51.23212988, last_update=datetime.datetime(2022, 10, 28, 0, 0, tzinfo=datetime.timezone.utc)),
    DisabledParking(entry_id='a7bc2fcf-072c-4b9f-8cf4-26f39890029e', number=1, address='Quirinstraße 32', district='Oberkassel', time_limit=None, note='vor der Kirche', longitude=6.75015157, latitude=51.23366626, last_update=datetime.datetime(2022, 10, 28, 0, 0, tzinfo=datetime.timezone.utc)),
  ])
# ---
# name: test_all_garages
  list([
    Garage(entry_id=1, name='Galeria-Kaufhof Am Wehrhahn', address='Oststraße 15', location='40211 Düsseldorf Stadtmitte', longitude=6.789029, latitude=51.227096),
    Garage(entry_id=2, name='Am Wehrhahn', address='Am Wehrhahn 72', location='40211 Düsseldorf Pempelfort', longitude=6.79381259809375, latitude=51.2292448423378),
    Garage(entry_id=3, name='Düsseldorf Arcaden', address='Bachstraße 141', location='40217 Düsseldorf Unterbilk', longitude=6.774042, latitude=51.209159),
    Garage(entry_id=4, name='Bendemannstraße', address='Bendemannstraße 5', location='40210 Düsseldorf Stadtmitte', longitude=6.79208848068309, latitude=51.2230525186179),
    Garage(entry_id=5, name='Carlslplatz', address='Benrather Straße 7B', location='40213 Düsseldorf Carlstadt', longitude=6.77366179353614, latitude=51.2229922455896),
    Garage(entry_id=6, name='Galeria Kaufhof Berliner Allee', address='Berliner Allee 52', location='40212 Düsseldorf Stadtmitte', longitude=6.78229566757991, latitude=51.219316243748),
    Garage(entry_id=7, name='Autohaus Moll', address='Berliner Allee 53', location='40211 Düsseldorf Stadtmitte', longitude=6.78173080974783, latitude=51.2204457432369),
    Garage(entry_id=8, name='Verkehrswacht Schadowstraße', address='Schadowstraße 76', location='40212 Düsseldorf Stadtmitte', longitude=6.785615, latitude=51.22788),
    Garage(entry_id=9, name='Carsch-Haus', address='Breite Straße 1', location='40213 Düsseldorf Stadtmitte', longitude=6.776906, latitude=51.224377),
    Garage(entry_id=10, name='E.On-Platz', address='Brüderstraße', location='40477 Düsseldorf Pempelfort', longitude=6.77303398, latitude=51.23565094),
  ])
# ---
# name: test_all_park_and_rides
  list([
    ParkAndRide(entry_id=1, name='Rath S', address='Westfalenstraße 90', district='06', neighbourhood='Rath', public_transport='Linie S6, U71, U72, 701', longitude=6.819546403, latitude=51.26421809),
    ParkAndRide(entry_id=2, name='Am Südfriedhof', address='Am Südfriedhof 11', district='03', neighbourhood='Bilk', public_transport='Linie 709, 723, 893', longitude=6.755868221, latitude=51.20038262),
    ParkAndRide(entry_id=3, name='Staufenplatz', address='Ludenberger Straße 8', district='07', neighbourhood='Grafenberg', public_transport='Linie M1, M2, U73, U83, 709,\xa0730, 733', longitude=6.829322103, latitude=51.24232978),
    ParkAndRide(entry_id=4, name='Flughafen Bahnhof', address='Ahrensplatz 9', district='06', neighbourhood='Lichtenbroich', public_transport='Linie IC/IR, RE1, RE2, RE3, RE5, RE6, RB35, S1, SB51, 729, 759, 776, SkyTrain', longitude=6.788162234, latitude=51.29058269),
    ParkAndRide(entry_id=5, name='Benrath S', address='Friedhofstraße 26', district='09', neighbourhood='Benrath', public_transport='Linie RE1, RE5, S6, S68, SB59, M1, U71, U74, U83,\xa0730, 778, 779, 784, 788, 789, 815, 817, NE7', longitude=6.87776289, latitude=51.16297837),
    ParkAndRide(entry_id=6, name='Eller S', address='Vennhauser Allee 89', district='08', neighbourhood='Eller', public_transport='Linie S1, M1, 722, 730, 735, 736, 891, NE6', longitude=6.852149131, latitude=51.20321557),
    ParkAndRide(entry_id=7, name='Angermund S', address='An den Linden 16', district='05', neighbourhood='Angermund', public_transport='Linie S1, 751', longitude=6.781381138, latitude=51.33036745),
    ParkAndRide(entry_id=8, name='Eller Süd S', address='Sturmstraße 80', district='08', neighbourhood='Eller', public_transport='Linie S6, S68, 724, 731, 732, 735, NE6', longitude=6.8363527, latitude=51.19845463),
    ParkAndRide(entry_id=9, name='Garath S (Ostseite u. Westseite)', address='Koblenzer Straße 253', district='10', neighbourhood='Garath', public_transport='Linie S6, S68, 778, 779, 789', longitude=6.896467257, latitude=51.14421404),
    ParkAndRide(entry_id=10, name='Hellerhof S', address='Carlo-Schmid-Straße 9', district='10', neighbourhood='Hellerhof', public_transport='Linie S6, S68, 789, 795', longitude=6.908925009, latitude=51.13145982),
  ])
# ---
